java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.ql

```diff
@@ -0,0 +1,153 @@
+/**
+ * @name Use of a hash function without a salt
+ * @description Hashed passwords without a salt are vulnerable to dictionary attacks.
+ * @kind path-problem
+ * @id java/hash-without-salt
+ * @tags security
+ *       external/cwe-759
+ */
+
+import java
+import semmle.code.java.dataflow.TaintTracking
+import semmle.code.java.dataflow.TaintTracking2
+import DataFlow::PathGraph
+
+/** The Java class `java.security.MessageDigest`. */
+class MessageDigest extends RefType {
+  MessageDigest() { this.hasQualifiedName("java.security", "MessageDigest") }
+}
+
+class MDConstructor extends StaticMethodAccess {
+  MDConstructor() {
+    exists(Method m | m = this.getMethod() |
+      m.getDeclaringType() instanceof MessageDigest and
+      m.hasName("getInstance")
+    )
+  }
+}
+
+/** The method `digest()` declared in `java.security.MessageDigest`. */
+class MDDigestMethod extends Method {
+  MDDigestMethod() {
+    this.getDeclaringType() instanceof MessageDigest and
+    this.hasName("digest")
+  }
+}
+
+/** The method `update()` declared in `java.security.MessageDigest`. */
+class MDUpdateMethod extends Method {
+  MDUpdateMethod() {
+    this.getDeclaringType() instanceof MessageDigest and
+    this.hasName("update")
+  }
+}
+
+/** The hashing method that could taint the input. */
+class MDHashMethodAccess extends MethodAccess {
+  MDHashMethodAccess() {
+    (
+      this.getMethod() instanceof MDDigestMethod or
+      this.getMethod() instanceof MDUpdateMethod
+    ) and
+    this.getNumArgument() != 0
+  }
+}
+
+/** Gets a regular expression for matching common names of variables that indicate the value being held is a password. */
+string getPasswordRegex() { result = "(?i).*pass(wd|word|code|phrase).*" }
+
+/** Finds variables that hold password information judging by their names. */
+class PasswordVarExpr extends Expr {
+  PasswordVarExpr() {
+    this.(VarAccess).getVariable().getName().toLowerCase().regexpMatch(getPasswordRegex()) and
+    not this.(VarAccess).getVariable().getName().toLowerCase().matches("%hash%") // Exclude variable names such as `passwordHash` since their values were already hashed
+  }
+}
+
+/** Holds if `Expr` e is a direct or indirect operand of `ae`. */
+predicate hasAddExpr(AddExpr ae, Expr e) { ae.getAnOperand+() = e }
+
+/** Holds if `MethodAccess` ma has a flow to another `MDHashMethodAccess` call. */
```

Comment:
```suggestion
/** Holds if `MDHashMethodAccess ma`'s qualifier or arguments flow to or from those of another `MDHashMethodAccess` call. */
```

Reply:
I've simplified the implementation since it only checks calls in the same enclosing block and updated the comment.